### ISSUE TITLE
NDI source stalls since timestamp changes

### ISSUE THREAD

**Issue by normen** [teltek/gst-plugin-ndi#79]
Hi,

Issue description:
When using a `ndisrc` as described in the README. the pipeline starts up but remains on time 0:00:00:000 forever, stalling the pipeline. In some random, rare cases, with uncompressed (non-HX) sources the pipeline starts to run.

After fighting with these stalling issues in the latest version for a while (both on MacOS and Windows) I found that the problem seems to exist since c98b626f69418e4cf796214e0fc7a527c5cb8364 - versions up to 7c3e50c629ca9c570502f73b996b8b816282dc1c seem to be working as expected.

Thanks for this plugin,
Normen

**Comment by sdroege** [comment]
How does your pipeline look like? Maybe a full `GST_DEBUG=9` would be useful here...

**Comment by normen** [comment]
Correct, I didn't see or hear anything - these logs are from the exact pipeline in the README using the official "NDI HX Camera" app on iPhone/Pad as a source.

Increasing the `max-size-time` on both queues to 10s didn't _really_ solve the issue - the video starts playing but after a few seconds I get the "dropped buffer" messages again, the audio is very choppy if its playing at all. I also get a lot of `Too small/big slope` errors..

**Comment by normen** [comment]
Funnily, in a different context ([obs-gstreamer plugin](https://github.com/fzwoch/obs-gstreamer)) the queue seems to work when using `timestamp-mode=2` - no additional buffers, no different settings, even when the appsink is synced...

However I noticed that the pipeline doesn't seem to react when setting an overall latency on the pipeline - this usually works for USB and RTSP sources to delay/sync them. The NDI pipeline always comes in with the same latency, no matter what I set for the latency, like this:

```
    gst_pipeline_set_latency (GST_PIPELINE (data->pipe), latency * GST_MSECOND);
    gint cur_latency = gst_pipeline_get_latency(GST_PIPELINE(data->pipe)) / GST_MSECOND;
    blog(LOG_INFO, "Set latency for pipeline to %d", cur_latency);
```

**Comment by sdroege** [comment]
> Except for the NDI pipeline which even with 10 seconds set as latency doesn't change its "real world" latency in the pipe.

As said, check the appsink/clock logs :) This is a different problem to what this issue is about though.

----

If you send me an iOS device, sure ;) I was testing with the NDI HX Camera app on Android with Linux as a receiver and that seems to be working correctly here.

**Comment by normen** [comment]
> > Except for the NDI pipeline which even with 10 seconds set as latency doesn't change its "real world" latency in the pipe.
> 
> As said, check the appsink/clock logs :) This is a different problem to what this issue is about though.
> 

Heres a level 6 log of OBS running with only one Gst-NDI source, I can't find any hints that setting the latency to 250ms fails, still its not applied in reality:

[obslog.txt](https://github.com/teltek/gst-plugin-ndi/files/7380806/obslog.txt)

**Comment by sdroege** [comment]
Like I said, everything's arriving too late in the sink

```
0:00:10.435205900  8364 000001E7B475C2C0 DEBUG               basesink gstbasesink.c:2785:gst_base_sink_do_sync:<video_appsink> clock returned 1, jitter  0:00:07.335453700
```

Means that this frame arrived 7.3s too late. As such, they will just be rendered immediately.

That might be a hint to the underlying problem here. Wrong timestamps in some sense.

**Comment by sdroege** [comment]
Wrong timestamping.

**Comment by sdroege** [comment]
If you check the logs, all the `NDIlib_audio_frame_v3_t` and `NDIlib_video_frame_v2_t` have a timecode of 0. So if you use any of the timecode-based modes it's not going to work very well. We should probably detect this case though.

If you select a timestamp or receive time based mode then it should work though. Can you give me a log for `timestamp-vs-receive-time` and another one for `receive-time`?

**Comment by normen** [comment]
Heres two logs, one with mode 0 and one with mode 4, neither yield an image in OBS, the previous was with mode 2, the only one that works somewhat.
[obslog-mode-4.txt](https://github.com/teltek/gst-plugin-ndi/files/7381530/obslog-mode-4.txt)
[obslog-mode-0.txt](https://github.com/teltek/gst-plugin-ndi/files/7381526/obslog-mode-0.txt)

**Comment by sdroege** [comment]
Thanks, I meant mode 1 (timestamp-vs...) and not 0 (timecode-vs...) though, can you also attach that? 0 is also useful however.

Mode 2 can't really work if the NDI sender is passing 0 for all timecodes. And generally (unless you really know what you're doing), you want either mode 0, 1 or 4 :)

**Comment by normen** [comment]
Right, sorry. Well if mode 2 is the only one that yields an image expect me to use that ;)
[obslog-mode-1.txt](https://github.com/teltek/gst-plugin-ndi/files/7382002/obslog-mode-1.txt)

**Comment by normen** [comment]
Okay, I don't get how thats necessary for internal receive times but here goes..
[obslog-mode-4-10sbuf.txt](https://github.com/teltek/gst-plugin-ndi/files/7387534/obslog-mode-4-10sbuf.txt)
[obslog-mode-1-10sbuf.txt](https://github.com/teltek/gst-plugin-ndi/files/7387533/obslog-mode-1-10sbuf.txt)
[obslog-mode-0-10sbuf.txt](https://github.com/teltek/gst-plugin-ndi/files/7387532/obslog-mode-0-10sbuf.txt)

**Comment by sdroege** [comment]
I finally got to look at these logs and trying to look into fixing the whole situation. I'll take your logs as one of my testcases for a more robust algorithm for the timestamps.

One thing that seems problematic in your log is
  1. timecodes are all set to 0 (OK, we should be able to handle that)
  2. timestamps between audio and video are apart 1.3s in the beginning, and audio timestamps are the same for all frames. This suggests a sender bug and we should try to detect this situation too somehow and fall back to using plain receive times here.

![Figure_1](https://user-images.githubusercontent.com/301846/171156323-79230d3a-1950-4260-8d63-3fbb540c9101.png)

**Comment by normen** [comment]
Okay, cool. The Sender is the official NDI HX Camera app for iOS in case it matters.

**Comment by sdroege** [comment]
Yeah I heard a few times already that it is apparently broken. The Android version of the app is behaving correctly.

**Comment by normen** [comment]
I guess it might also be a matter with the NDI core libraries on the "darwin" platforms.. I also get strange behavior when I try to display a NDI screen grab from an iMac.

**Comment by sdroege** [comment]
This should be fixed by https://github.com/teltek/gst-plugin-ndi/pull/102 with the default settings now

**Comment by normen** [comment]
Cool, it'll be some time until I can check but I'll report back!

**Comment by normen** [comment]
So, coming back to testing this and on sources that have *no* audio the latest version works a treat. For sources *with* audio not so much, I couldn't get it to work with actual NDI hardware, or the NDI HX iPhone app or with the new NDICam App or even with a screen capture session. It always gets stuck at 00:00:00 and seems to drop pretty much all frames.

Heres a log with the new NDICam App:
[log.txt](https://github.com/teltek/gst-plugin-ndi/files/9719343/log.txt)

Edit: Also, I get errors when I don't grab the audio from the demuxer. I didn't know this kind of behavior from gstreamer, usually when I don't connect an audio sink the audio frames simply get dropped (e.g. RTSP).

**Comment by normen** [comment]
D'oh, I just realized your PR wasn't merged yet, checked out your branch now and it works great, thanks a lot!

**Comment by sdroege** [comment]
Thanks for testing :)

**Comment by sdroege** [comment]
I've integrated the fix here now: https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/tree/main/net/ndi

**Comment by normen** [comment]
Hmmm.. It looks like I have to come back to this. With the latest version from the gst-plugins-rs I still get the described issues. Every time audio gets in the mix the stream stalls. When I have a source without audio everything works fine. One issue is that I can't drop the audio packets in a pipeline that has audio, I have to pick up the audio pad or I get a pipeline error. Usually in gstreamer I don't have to do that?

Also I get mixed up with all the versions.. It seems all versions that have the `ndisrcdemux` element show this behavior. The version from your private repo that has separate `ndivideosrc` and `ndiaudiosrc` elements works best for me. I am currently using that to stream my NDI cameras.

**Comment by sdroege** [comment]
Can you report that over there with the exact pipeline you're using and a `GST_DEBUG=6,ndi*:9` debug log? That should help finding the cause. Like this there's no actionable information :)
